File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Panels/PlayerStatsPanel.cs
using ScaleformUI.Elements;
using ScaleformUI.LobbyMenu;
using ScaleformUI.Menu;
using ScaleformUI.Menus;
using ScaleformUI.PauseMenu;
using ScaleformUI.PauseMenus.Elements.Items;
using ScaleformUI.Scaleforms;
using System.Collections.Generic;
using static CitizenFX.Core.Native.API;

namespace ScaleformUI.PauseMenus.Elements.Panels
{
    public class PlayerStatsPanel : PM_Column
    {
        public FriendItem ParentItem { get; set; }
        private string title;
        private ScaleformLabel description = "";
        private SColor titleColor = SColor.FromHudColor(HudColor.HUD_COLOUR_FREEMODE);
        private bool hasPlane = false;
        private bool hasVehicle = false;
        private bool hasBoat = false;
        private bool hasHeli = false;

        public UpperInformation RankInfo { get; set; }
        public BottomInformation CrewInfo { get; set; }

        public string Title
        {
            get => title;
            set
            {
                title = value;
                if (visible)
                    UpdatePanel();
            }
        }
        public SColor TitleColor
        {
            get => titleColor;
            set
            {
                titleColor = value;
                if (visible)
                    UpdatePanel();
            }
        }
        public ScaleformLabel Description
        {
            get => description;
            set
            {
                description = value;
                if (visible)
                    UpdatePanel();
            }
        }
        public bool HasPlane
        {
            get => hasPlane;
            set
            {
                hasPlane = value;
                if (visible)
                    UpdatePanel();
            }
        }
        public bool HasHeli
        {
            get => hasHeli;
            set
            {
                hasHeli = value;
                if (visible)
                    UpdatePanel();
            }
        }
        public bool HasBoat
        {
            get => hasBoat;
            set
            {
                hasBoat = value;
                if(visible)
                    UpdatePanel();
            }
        }
        public bool HasVehicle
        {
            get => hasVehicle;
            set
            {
                hasVehicle = value;
                if(visible)
                    UpdatePanel();
            }
        }
        public bool HardwareVisible { get; set; } = true;
        public List<UIFreemodeDetailsItem> DetailsItems { get; private set; }
        public PlayerStatsPanel(string title, SColor titleColor) : base(PM_COLUMNS.RIGHT)
        {
            VisibleItems = 10;
            this.title = title;
            this.titleColor = titleColor;
            this.description = string.Empty;
            RankInfo = new(this);
            CrewInfo = new(this);
            Items = new();
            DetailsItems = new();
        }

        public void AddStat(PlayerStatsPanelStatItem item)
        {
            item.Parent = this;
            Items.Add(item);
            if(visible)
                UpdatePanel();
        }

        public void AddDescriptionStatItem(UIFreemodeDetailsItem item)
        {
            DetailsItems.Add(item);
            if(visible)
                UpdatePanel();
        }

        public void UpdatePanel(bool _override = false)
        {
            if (ParentItem != null && ParentItem.ParentColumn != null && ParentItem.ParentColumn.visible || _override)
            {
                position = ParentItem?.ClonePed == null ? PM_COLUMNS.EXTRA3 : PM_COLUMNS.EXTRA4;
                Parent = ParentItem?.ParentColumn?.Parent;
                Populate();
                ShowColumn();
            }
        }

        public override void Populate()
        {
            Main.PauseMenu._pause.CallFunction("SET_DATA_SLOT_EMPTY", (int)position);
            if(CrewInfo.isFilled)
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_TITLE", (int)position, Title, CrewInfo.CrewName, TitleColor, "", CrewInfo.RankName, ParentItem.CrewTag != null ? ParentItem.CrewTag.TAG : "", CrewInfo.CrewDict, CrewInfo.CrewTxtr, CrewInfo.CrewTag, CrewInfo.CrewColor.R, CrewInfo.CrewColor.G, CrewInfo.CrewColor.B, 0, "");
            else
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_TITLE", (int)position, Title, "", TitleColor, "", "", ParentItem.CrewTag != null ? ParentItem.CrewTag.TAG : "", "", "", "", 0, 0, 0, 0, "");
            Main.PauseMenu._pause.CallFunction("SET_DATA_SLOT", (int)position, 0, 0, 0, 0, HasPlane, HasHeli, HasBoat, HasVehicle, 0, RankInfo.RankLevel, RankInfo.LowLabel, 0, 0, RankInfo.MidLabel, 0, 0, RankInfo.UpLabel, 0, 0, HardwareVisible);
            for (int i = 0; i < Items.Count; i++)
            {
                PlayerStatsPanelStatItem stat = (PlayerStatsPanelStatItem)Items[i];
                Main.PauseMenu._pause.CallFunction("SET_DATA_SLOT", (int)position, 1, i, 0, stat.Label, stat.Description, stat.Value);
            }
            if (DetailsItems.Count > 0 && position == PM_COLUMNS.EXTRA3)
            {
                for (int i = 0; i < DetailsItems.Count; i++)
                {
                    UIFreemodeDetailsItem item = DetailsItems[i];
                    BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DATA_SLOT");
                    PushScaleformMovieFunctionParameterInt((int)position);
                    PushScaleformMovieFunctionParameterInt(2);
                    PushScaleformMovieFunctionParameterInt(i);
                    PushScaleformMovieFunctionParameterInt(0);
                    PushScaleformMovieFunctionParameterInt(0);
                    PushScaleformMovieFunctionParameterInt(item.Type);
                    PushScaleformMovieFunctionParameterInt(0);
                    PushScaleformMovieFunctionParameterBool(true);
                    var labels = item.Label.SplitLabel;
                    BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                    for (var j = 0; j < labels?.Length; j++)
                        AddTextComponentScaleform(labels[j]);
                    EndTextCommandScaleformString_2();
                    PushScaleformMovieFunctionParameterString(item.TextRight);
                    switch (item.Type)
                    {
                        case 2:
                            PushScaleformMovieFunctionParameterInt((int)item.Icon);
                            PushScaleformMovieFunctionParameterInt(item.IconColor.ArgbValue);
                            PushScaleformMovieFunctionParameterBool(item.Tick);
                            break;
                        case 3:
                            PushScaleformMovieFunctionParameterString(item.CrewTag.TAG);
                            PushScaleformMovieFunctionParameterBool(false);
                            break;
                    }
                    PushScaleformMovieFunctionParameterString(item.LabelFont.FontName);
                    PushScaleformMovieFunctionParameterString(item._rightLabelFont.FontName);
                    EndScaleformMovieMethod();
                }
            }
            else if (!string.IsNullOrWhiteSpace(Description.Label) && !CrewInfo.isFilled)
            {
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_DESCRIPTION");
                PushScaleformMovieFunctionParameterInt((int)position);
                var labels = description.SplitLabel;
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                for (var j = 0; j < labels?.Length; j++)
                    AddTextComponentScaleform(labels[j]);
                EndTextCommandScaleformString_2();
                PushScaleformMovieFunctionParameterInt(0);
                PushScaleformMovieFunctionParameterString(ParentItem.CrewTag != null ? ParentItem.CrewTag.TAG : "");//crewtag?
                PushScaleformMovieFunctionParameterBool(ParentItem.ClonePed != null);
                EndScaleformMovieMethod();

            }
        }

        public override void SetDataSlot(int index)
        {
            Populate();
        }


        public override void UpdateSlot(int index)
        {
            if (index >= Items.Count) return;
            if (visible)
                Populate();
        }

        public override void AddSlot(int index)
        {
            if (index >= Items.Count) return;
            if (visible)
                Populate();
        }
    }
    public sealed class UpperInformation
    {
        internal PlayerStatsPanel _parent;
        private int rankLevel = 0;
        private string upLabel = "";
        private string lowLabel = "";
        private string midLabel = "";

        public int RankLevel
        {
            get => rankLevel;
            set
            {
                rankLevel = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string UpLabel
        {
            get => upLabel;
            set
            {
                upLabel = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string MidLabel
        {
            get => midLabel;
            set
            {
                midLabel = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string LowLabel
        {
            get => lowLabel;
            set
            {
                lowLabel = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public UpperInformation(PlayerStatsPanel parent)
        {
            _parent = parent;
        }
    }

    public sealed class BottomInformation
    {
        internal bool isFilled => !string.IsNullOrWhiteSpace(CrewName) && !string.IsNullOrWhiteSpace(RankName) && !string.IsNullOrWhiteSpace(CrewDict) && !string.IsNullOrWhiteSpace(CrewTxtr) && !string.IsNullOrWhiteSpace(CrewTag);
        private PlayerStatsPanel _parent;
        private string crewName = "";
        private string rankName = "";
        private string crewDict = "";
        private string crewTxtr = "";
        private string crewtag = "";
        private SColor crewColor = SColor.HUD_Pure_white;

        public string CrewName
        {
            get => crewName;
            set
            {
                crewName = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string RankName
        {
            get => rankName;
            set
            {
                rankName = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string CrewDict
        {
            get => crewDict;
            set
            {
                crewDict = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string CrewTxtr
        {
            get => crewTxtr;
            set
            {
                crewTxtr = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public SColor CrewColor
        {
            get => crewColor;
            set
            {
                crewColor = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }
        public string CrewTag
        {
            get => crewtag;
            set
            {
                crewtag = value;
                if (_parent.visible)
                    _parent.UpdatePanel();
            }
        }

        public BottomInformation(PlayerStatsPanel parent)
        {
            _parent = parent;
        }
    }

    public sealed class PlayerStatsPanelStatItem : PauseMenuItem
    {

        internal int idx { get; set; }
        private int _value;
        private string description;
        private string label;
        public new PlayerStatsPanel Parent { get; internal set; }
        public new string Label
        {
            get => label;
            set
            {
                base.Label = value;
                label = value;
                Parent.UpdatePanel();
            }
        }
        public string Description
        {
            get => description;
            set
            {
                description = value;
                Parent.UpdatePanel();
            }
        }
        public int Value
        {
            get => _value;
            set
            {
                _value = value;
                Parent.UpdatePanel();
            }
        }

        public PlayerStatsPanelStatItem(string label, string description, int value) : base(label)
        {
            this.label = label;
            this.description = description;
            this._value = value;
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Items/ItemPanels/PlayerStatsPanel.lua
PlayerStatsPanel = {}
PlayerStatsPanel.__index = PlayerStatsPanel
setmetatable(PlayerStatsPanel, { __index = PM_Column })
PlayerStatsPanel.__call = function() return "PlayerStatsPanel" end

---@class PlayerStatsPanel
---@field private _title string
---@field private _description string
---@field private _titleColor number
---@field private _hasPlane boolean
---@field private _hasVehicle boolean
---@field private _hasBoat boolean
---@field private _hasHeli boolean
---@field public ParentItem FriendItem
---@field public RankInfo UpperInformation
---@field public Items PlayerStatsPanelStatItem[]
---@field public UpdatePanel fun(self: PlayerStatsPanel, override: boolean?)
---@field public Description fun(self: PlayerStatsPanel, desc: string?): string
---@field public HasPlane fun(self: PlayerStatsPanel, bool: boolean?): boolean
---@field public HasHeli fun(self: PlayerStatsPanel, bool: boolean?): boolean
---@field public HasBoat fun(self: PlayerStatsPanel, bool: boolean?): boolean
---@field public HasVehicle fun(self: PlayerStatsPanel, bool: boolean?): boolean
---@field public HardwareVisible fun(self: PlayerStatsPanel, bool: boolean?): boolean
---@field public AddStat fun(self: PlayerStatsPanel, statItem: PlayerStatsPanelStatItem)
---@field public OnItemChanged fun(item: PlayerStatsPanelStatItem)
---@field public OnItemActivated fun(item: PlayerStatsPanelStatItem)

---Creates a new PlayerStatsPanel.
---@param title string
---@param titleColor number
---@return PlayerStatsPanel
function PlayerStatsPanel.New(title, titleColor)
    local base = PM_Column.New(3)
    base.VisibleItems = 10
    base.ParentItem = nil
    base._hardwareVisible = true
    base._title = title or ""
    base._description = ""
    base._titleColor = titleColor or SColor.HUD_Freemode
    base._hasPlane = false
    base._hasVehicle = false
    base._hasBoat = false
    base._hasHeli = false
    base.RankInfo = nil
    base.DetailsItems = {}
    local retVal = setmetatable(base, PlayerStatsPanel)
    retVal.RankInfo = UpperInformation.New(retVal)
    retVal.CrewInfo = BottomInformation.New(retVal)
    return retVal
end

function PlayerStatsPanel:visible()
    return PM_Column.visible(self)
end

function PlayerStatsPanel:HardwareVisible(v)
    if v == nil then
        return self._hardwareVisible
    else
        self._hardwareVisible = v
    end
end

---Sets the title of the panel if supplied else it will return the current title.
---@param label string?
---@return string
function PlayerStatsPanel:Title(label)
    if label ~= nil then
        self._title = label
        self:UpdatePanel()
    end
    return self._title
end

---Sets the title color of the panel if supplied else it will return the current color.
---@param color SColor?
---@return SColor
function PlayerStatsPanel:TitleColor(color)
    if color ~= nil then
        self._titleColor = color
        self:UpdatePanel()
    end
    return self._titleColor
end

---Sets the description of the panel if supplied else it will return the current description.
---@param label string?
---@return string
function PlayerStatsPanel:Description(label)
    if label ~= nil then
        self._description = label
        self:UpdatePanel()
    end
    return self._description
end

---Sets whether the player has a plane or not, if parameter is nill, it will return the current value.
---@param bool boolean?
---@return boolean
function PlayerStatsPanel:HasPlane(bool)
    if bool ~= nil then
        self._hasPlane = bool
        self:UpdatePanel()
    end
    return self._hasPlane
end

---Sets whether the player has a helicopter or not, if parameter is nill, it will return the current value.
---@param bool boolean?
---@return boolean
function PlayerStatsPanel:HasHeli(bool)
    if bool ~= nil then
        self._hasHeli = bool
        self:UpdatePanel()
    end
    return self._hasHeli
end

---Sets whether the player has a boat or not, if parameter is nill, it will return the current value.
---@param bool boolean?
---@return boolean
function PlayerStatsPanel:HasBoat(bool)
    if bool ~= nil then
        self._hasBoat = bool
        self:UpdatePanel()
    end
    return self._hasBoat
end

---Sets whether the player has a vehicle or not, if parameter is nill, it will return the current value.
---@param bool boolean?
---@return boolean
function PlayerStatsPanel:HasVehicle(bool)
    if bool ~= nil then
        self._hasVehicle = bool
        self:UpdatePanel()
    end
    return self._hasVehicle
end

---Adds a new stat item to the panel.
---@param statItem PlayerStatsPanelStatItem
function PlayerStatsPanel:AddStat(statItem)
    statItem.Parent = self
    statItem._idx = #self.Items
    table.insert(self.Items, statItem)
    if self:visible() then
        self:UpdatePanel()
    end
end

function PlayerStatsPanel:AddDescriptionStatItem(item)
    table.insert(self.DetailsItems, item)
    if self:visible() then
        self:UpdatePanel()
    end
end

---Triggers the panel to update.
---@param override boolean? If true, the panel will update regardless of the parent's visibility.
function PlayerStatsPanel:UpdatePanel(override)
    if ((self.ParentItem ~= nil and self.ParentItem.ParentColumn ~= nil and self.ParentItem.ParentColumn:visible()) or override) then
        if self.ParentItem.ClonePed == 0 then
            self.position = 3
        else
            self.position = 4
        end
        self.Parent = self.ParentItem.ParentColumn.Parent
        self:Populate()
        self:ShowColumn()
    end
end

function PlayerStatsPanel:Populate()
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT_EMPTY", self.position)
    if self.CrewInfo:IsFilled() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_TITLE", self.position, self._title,
            self.CrewInfo:CrewName(), self:TitleColor(), "", self.CrewInfo:RankName(), ParentItem.CrewTag,
            self.CrewInfo:CrewDict(), self.CrewInfo:CrewTxtr(), self.CrewInfo:CrewTag().TAG, self.CrewInfo:CrewColor().R,
            self.CrewInfo:CrewColor().G, self.CrewInfo:CrewColor().B, 0, "")
    else
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_TITLE", self.position, self._title, "",
            self:TitleColor(), "", "", self.ParentItem:CrewTag().TAG, "", "", "", 0, 0, 0, 0, "")
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT", self.position, 0, 0, 0, 0, self:HasPlane(),
        self:HasHeli(), self:HasBoat(), self:HasVehicle(), 0, self.RankInfo:RankLevel(), self.RankInfo:LowLabel(), 0, 0,
        self.RankInfo:MidLabel(), 0, 0, self.RankInfo:UpLabel(), 0, 0, self:HardwareVisible())
    for i = 1, #self.Items, 1 do
        local stat = self.Items[i]
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT", self.position, 1, i - 1, 0, stat:Label(),
            stat:Description(), stat:Value())
    end
    if #self.DetailsItems > 0 and self.position == 3 then
        for i = 1, #self.DetailsItems, 1 do
            local item = self.DetailsItems[i]
            BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DATA_SLOT")
            PushScaleformMovieFunctionParameterInt(self.position)
            PushScaleformMovieFunctionParameterInt(2)
            PushScaleformMovieFunctionParameterInt(i - 1)
            PushScaleformMovieFunctionParameterInt(0)
            PushScaleformMovieFunctionParameterInt(0)
            PushScaleformMovieFunctionParameterInt(item.Type)
            PushScaleformMovieFunctionParameterInt(0)
            PushScaleformMovieFunctionParameterBool(true)
            local labels = item.label:SplitLabel()
            BeginTextCommandScaleformString("CELL_EMAIL_BCON")
            for j = 1, #labels, 1 do
                AddTextComponentScaleform(labels[j])
            end
            EndTextCommandScaleformString_2()
            PushScaleformMovieFunctionParameterString(item.TextRight)
            if item.Type == 2 then
                PushScaleformMovieFunctionParameterInt(item.Icon)
                PushScaleformMovieFunctionParameterInt(item.IconColor:ToArgb())
                PushScaleformMovieFunctionParameterBool(item.Tick)
            elseif item.Type == 3 then
                PushScaleformMovieFunctionParameterString(item.CrewTag.TAG)
                PushScaleformMovieFunctionParameterBool(false)
            end
            PushScaleformMovieFunctionParameterString(item.LabelFont.FontName)
            PushScaleformMovieFunctionParameterString(item._rightLabelFont.FontName)
            EndScaleformMovieMethod()
        end
    elseif not self:Description():IsNullOrEmpty() and not self.CrewInfo:IsFilled() then
        BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_DESCRIPTION")
        PushScaleformMovieFunctionParameterInt(self.position)
        local labels = self:Description():SplitLabel()
        BeginTextCommandScaleformString("CELL_EMAIL_BCON")
        for j = 1, #labels, 1 do
            AddTextComponentScaleform(labels[j])
        end
        EndTextCommandScaleformString_2()
        PushScaleformMovieFunctionParameterInt(0)
        PushScaleformMovieFunctionParameterString(self.ParentItem:CrewTag().TAG)
        PushScaleformMovieFunctionParameterBool(ToBool(self.ParentItem.ClonePed ~= 0))
        EndScaleformMovieMethod()
    end
end

function PlayerStatsPanel:SetDataSlot(index)
    self:Populate()
end

function PlayerStatsPanel:UpdateSlot(index)
    if index > #self.Items then return end
    if self:visible() then
        self:Populate()
    end
end
function PlayerStatsPanel:AddSlot(index)
    if index > #self.Items then return end
    if self:visible() then
        self:Populate()
    end
end